pkg/log/log.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package log
+
+import (
+	"io"
+	"strconv"
+	"sync"
+	"sync/atomic"
+)
+
+// Level specifies a level of verbosity for logger.
+// Level should be modified only through its set method.
+// Level is treated as a sync/atomic int32.
+type Level int32
+
+// get returns the value of the Level.
+func (l *Level) get() Level {
+	return Level(atomic.LoadInt32((*int32)(l)))
+}
+
+// set updates the value of the Level.
+func (l *Level) set(v Level) {
+	atomic.StoreInt32((*int32)(l), int32(v))
+}
+
+// String implements the fmt.Stringer interface.
+func (l Level) String() string {
+	switch l.get() {
+	case VerbosityNone:
+		return "none"
+	case VerbosityError:
+		return "error"
+	case VerbosityWarning:
+		return "warning"
+	case VerbosityInfo:
+		return "info"
+	case VerbosityDebug:
+		return "debug"
+	case VerbosityAll:
+		return "all"
+	}
+	return strconv.FormatInt(int64(l), 10) // Covers all in the range [VerbosityDebug ... VerbosityAll>.
+}
+
+// ParseVerbosityLevel returns a verbosity Level parsed from the given s.
+func ParseVerbosityLevel(s string) (Level, error) {
+	switch s {
+	case "none":
+		return VerbosityNone, nil
+	case "error":
+		return VerbosityError, nil
+	case "warning":
+		return VerbosityWarning, nil
+	case "info":
+		return VerbosityInfo, nil
+	case "debug":
+		return VerbosityDebug, nil
+	case "all":
+		return VerbosityAll, nil
+	}
+	i, err := strconv.ParseInt(s, 10, 32)
+	return Level(i), err
+}
+
+const (
+	// VerbosityNone will silence the logger.
+	VerbosityNone = Level(iota - 4)
+	// VerbosityError allows only error messages to be printed.
+	VerbosityError
+	// VerbosityWarning allows only error and warning messages to be printed.
+	VerbosityWarning
+	// VerbosityInfo allows only error, warning and info messages to be printed.
+	VerbosityInfo
+	// VerbosityDebug allows only error, warning, info and debug messages to be printed.
+	VerbosityDebug
+	// VerbosityAll allows to print all messages up to and including level V.
+	// It is a placeholder for the maximum possible V level verbosity within the logger.
+	VerbosityAll = Level(1<<31 - 1)
+)
+
+// Hook that is fired when logging
+// on the associated severity log level.
+// Note, the call must be non-blocking.
```

Comment:
So although I see that there could be some use for this, I would reconsider having this hook functionality. This could be easily abused, given that this is called during the happy path. I think with the options and the builder we have enough ways to configure things.